File: gatelogue-types-rs/src/node/bus.rs
use strum_macros::EnumString;

use crate::{_from_sql_for_enum, _get_column, _get_derived_vec, _get_set, node_type, util::ID};

#[derive(Clone, Copy, PartialEq, Eq, Debug, EnumString)]
pub enum BusMode {
    #[strum(serialize = "warp")]
    Warp,
    #[strum(serialize = "traincarts")]
    TrainCarts,
}
_from_sql_for_enum!(BusMode);

node_type!(BusCompany);
impl BusCompany {
    _get_column!("BusCompany", name, String);
    _get_column!("BusCompany", link, Option<String>);
    _get_derived_vec!(lines, BusLine, "../sql/bus/company_lines.sql");
    _get_derived_vec!(stops, BusStop, "../sql/bus/company_stops.sql");
    _get_derived_vec!(berths, BusBerth, "../sql/bus/company_berths.sql");
}

node_type!(BusLine);
impl BusLine {
    _get_column!("BusLine", code, String);
    _get_column!("BusLine", company, BusCompany);
    _get_column!("BusLine", name, Option<String>);
    _get_column!("BusLine", colour, Option<String>);
    _get_column!("BusLine", mode, Option<BusMode>);
    _get_column!("BusLine", local, Option<bool>);

    _get_derived_vec!(berths, BusBerth, "../sql/bus/line_berths.sql");
    _get_derived_vec!(stops, BusStop, "../sql/bus/company_stops.sql");
}

node_type!(located BusStop);
impl BusStop {
    _get_set!("BusStopCodes", codes, "code", String);
    _get_column!("BusStop", company, BusCompany);
    _get_column!("BusStop", name, Option<String>);

    _get_derived_vec!(berths, BusBerth, "../sql/bus/stop_berths.sql");
    _get_derived_vec!(
        connections_from_here,
        BusConnection,
        "../sql/bus/stop_connections_from_here.sql"
    );
    _get_derived_vec!(
        connections_to_here,
        BusConnection,
        "../sql/bus/stop_connections_to_here.sql"
    );
    _get_derived_vec!(lines, BusLine, "../sql/bus/stop_lines.sql");
}

node_type!(BusBerth);
impl BusBerth {
    _get_column!("BusBerth", code, Option<String>);
    _get_column!("BusBerth", stop, BusStop);

    _get_derived_vec!(
        connections_from_here,
        BusConnection,
        "../sql/bus/berth_connections_from_here.sql"
    );
    _get_derived_vec!(
        connections_to_here,
        BusConnection,
        "../sql/bus/berth_connections_to_here.sql"
    );
    _get_derived_vec!(lines, BusLine, "../sql/bus/berth_lines.sql");
}

node_type!(BusConnection);
impl BusConnection {
    _get_column!("BusConnection", line, BusLine);
    _get_column!("BusConnection", from, BusBerth);
    _get_column!("BusConnection", to, BusBerth);
    _get_column!("BusConnection", direction, Option<String>);
    _get_column!("BusConnection", duration, Option<u32>);
}

File: gatelogue-types-ts/src/bus.ts
import { Node } from "./node.js";
import { LocatedNode } from "./located.js";

export type BusMode = "warp" | "traincarts";

export class BusCompany extends Node {
  get name(): string {
    return this.getColumn("BusCompany", "name");
  }
  get link(): string | null {
    return this.getColumn("BusCompany", "link");
  }

  get lines(): BusLine[] {
    return this.getDerived(BusLine, "bus/company_lines");
  }
  get stops(): BusStop[] {
    return this.getDerived(BusStop, "bus/company_stops");
  }
  get berths(): BusBerth[] {
    return this.getDerived(BusBerth, "bus/company_berths");
  }
}

export class BusLine extends Node {
  get code(): string {
    return this.getColumn("BusLine", "code");
  }
  get company(): BusCompany {
    return this.getColumnFK("BusLine", "company", BusCompany)!;
  }
  get name(): string | null {
    return this.getColumn("BusLine", "name");
  }
  get colour(): string | null {
    return this.getColumn("BusLine", "colour");
  }
  get mode(): BusMode | null {
    return this.getColumn("BusLine", "name");
  }
  get local(): boolean | null {
    const result = this.getColumn<number>("BusLine", "name");
    return result === null ? null : result === 1;
  }

  get berths(): BusBerth[] {
    return this.getDerived(BusBerth, "bus/line_berths");
  }
  get stops(): BusStop[] {
    return this.getDerived(BusStop, "bus/line_stops");
  }
}

export class BusStop extends LocatedNode {
  get codes(): string[] {
    return this.getSet("BusStopCodes", "code");
  }
  get company(): BusCompany {
    return this.getColumnFK("BusStop", "company", BusCompany)!;
  }
  get name(): string | null {
    return this.getColumn("BusStop", "name");
  }

  get berths(): BusBerth[] {
    return this.getDerived(BusBerth, "bus/stop_berths");
  }
  get connectionsFromHere(): BusConnection[] {
    return this.getDerived(BusConnection, "bus/stop_connections_from_here");
  }
  get connectionsToHere(): BusConnection[] {
    return this.getDerived(BusConnection, "bus/stop_connections_to_here");
  }
  get lines(): BusLine[] {
    return this.getDerived(BusLine, "bus/stop_lines");
  }
}

export class BusBerth extends Node {
  get code(): string | null {
    return this.getColumn("BusBerth", "code");
  }
  get stop(): BusStop {
    return this.getColumnFK("BusBerth", "stop", BusStop)!;
  }

  get connectionsFromHere(): BusConnection[] {
    return this.getDerived(BusConnection, "bus/berth_connections_from_here");
  }
  get connectionsToHere(): BusConnection[] {
    return this.getDerived(BusConnection, "bus/berth_connections_to_here");
  }
  get lines(): BusLine[] {
    return this.getDerived(BusLine, "bus/berth_lines");
  }
}

export class BusConnection extends Node {
  get line(): BusLine {
    return this.getColumnFK("BusConnection", "line", BusLine)!;
  }
  get from(): BusBerth {
    return this.getColumnFK("BusConnection", "from", BusBerth)!;
  }
  get to(): BusBerth {
    return this.getColumnFK("BusConnection", "to", BusBerth)!;
  }
  get direction(): string | null {
    return this.getColumn("BusConnection", "direction");
  }
  get duration(): number | null {
    return this.getColumn("BusConnection", "duration");
  }
}
